extensions/amp-a4a/0.1/refresh-manager.js

```diff
@@ -0,0 +1,216 @@
+  ValidAdContainerTypes,
+} from '../../../ads/google/a4a/utils';
+import {user} from '../../../src/log';
+
+/**
+ * - visibilePercentageMin: The percentage of pixels that need to be on screen
+ *   for the creative to be considered "visible".
+ * - totalTimeMin: The total amount of time, in milliseconds, that the creative
+ *   must be on screen in order to be considered "visible".
+ * - continuousTimeMin: The amount of continuous time, in milliseconds, that
+ *   the creative must be on screen for in oreder to be considered "visible".
+ *
+ * @typedef {{
+ *   visiblePercentageMin: number,
+ *   totalTimeMin: number,
+ *   continuousTimeMin: number,
+ * }}
+ */
+export let RefreshConfig;
+
+export const MIN_REFRESH_INTERVAL = 30;
+export const DATA_ATTR_NAME = 'data-enable-refresh';
+export const METATAG_NAME = 'amp-ad-enable-refresh';
+
+const TAG = 'AMP-AD';
+
+export class RefreshManager {
+
+  /**
+   * @param {!./amp-a4a.AmpA4A} a4a The AmpA4A instance to be refreshed.
+   * @param {!RefreshConfig} config
+   */
+  constructor(a4a, config) {
+
+    /** @const @private {!./amp-a4a.AmpA4A} */
+    this.a4a_ = a4a;
+
+    /** @const @private {!Window} */
+    this.win_ = a4a.win;
+
+    /** @const @private {!Element} */
+    this.element_ = a4a.element;
+
+    /** @const @private {string} */
+    this.adType_ = this.element_.getAttribute('type').toLowerCase();
+
+    /** @const @private {?number} */
+    this.refreshInterval_ = this.getPublisherSpecifiedRefreshInterval_();
+
+    /** @const @private {!RefreshConfig} */
+    this.config_ = this.convertConfiguration_(config);
+
+    /** @const @private {!../../../src/service/timer-impl.Timer} */
+    this.timer_ = timerFor(this.win_);
+
+    /** @private {?(number|string)} */
+    this.refreshTimeoutId_ = null;
+
+    /** @private {boolean} */
+    this.isRefreshable_ = isExperimentOn(this.win_, 'amp-ad-refresh') &&
+        // The network has opted-in.
+        !!(this.config_
+        // The publisher has enabled refresh on this slot.
+        && (this.refreshInterval_ || this.refreshInterval_ == '')
+        // The slot is contained only within container types eligible for
+        // refresh.
+        && !getEnclosingContainerTypes(this.element_).filter(container =>
+          container != ValidAdContainerTypes['AMP-CAROUSEL']
+          && container != ValidAdContainerTypes['AMP-STICKY-AD']).length);
+
+    if (this.isRefreshable_) {
+      this.initiateRefreshCycle();
+    }
+  }
+
+
+  /**
+   * Initiates the refresh cycle by initiating the visibility manager on the
+   * element.
+   *
+   * @return {!Promise<boolean>} Promise that resolves to true if refresh event
+   *    completes successfully and false otherwise. This is particularly useful
+   *    for testing.
+   */
+  initiateRefreshCycle() {
+    if (!this.isRefreshable()) {
+      return Promise.resolve(false);
+    }
+    return new Promise(resolve => {
+      analyticsForDoc(this.element_, true).then(analytics => {
```

Comment:
@jridgewell since you're planning to provide visibility info in a central place with your layoutmanager work, for an easier upgrade to that, do you think here we should use InOb directly, or should we rely on the viewportCallback?